### PR TITLE
Add Kopia s3 and s3 compliant storage flags

### PR DESCRIPTION
**Pull request by plar**
## Change Overview

This PR is the 7th in the series of PRs that will add a new way to build Kopia CLI commands.   

PR Train:

1. [PR-2653](https://github.com/kanisterio/kanister/pull/2653) Introducing a new approach to building Kopia CLI commands using safecli v0.0.6
1. [PR-2654](https://github.com/kanisterio/kanister/pull/2654) Implement Kopia common args and opts
1. [PR-2655](https://github.com/kanisterio/kanister/pull/2655) Implement Kopia storage helpers
1. [PR-2656](https://github.com/kanisterio/kanister/pull/2656) Add Kopia Filesystem storage opts
1. [PR-2657](https://github.com/kanisterio/kanister/pull/2657) Add Kopia GCS storage opts
1. [PR-2658](https://github.com/kanisterio/kanister/pull/2658) Add Kopia Azure storage opts
1. **[PR-2659](https://github.com/kanisterio/kanister/pull/2659) Add Kopia S3 and S3 compliant storage opts**
1. [PR-2661](https://github.com/kanisterio/kanister/pull/2661) Add Kopia repository create command
1. [PR-2662](https://github.com/kanisterio/kanister/pull/2662) Add Kopia repository connect command
1. [PR-2663](https://github.com/kanisterio/kanister/pull/2663) Add Kopia repository connect server command
1. [PR-2664](https://github.com/kanisterio/kanister/pull/2664) Add Kopia repository set-parameters command
1. [PR-2665](https://github.com/kanisterio/kanister/pull/2665) Add Kopia repository status command

This PR introduces storage *s3* and *s3 compliant* flag structures which will be used to build storage flags for s3 and s3 compliant storages.

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [x] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

**Review by hairyhum — APPROVED** [review]

File: pkg/kopia/cli/internal/log/log.go [hunk 1]
```diff
@@ -43,3 +44,31 @@ func (NopLogger) WithContext(ctx context.Context) log.Logger {
...
+}
+
+func (l *StringLogger) MatchString(pattern string) bool {
+	for _, line := range *l {
+		if found, _ := regexp.MatchString(pattern, line); found {
+			return true
+		}
+	}
+	return false
+}
```
Comment (kale-amruta) [inline pkg/kopia/cli/internal/log/log.go:74]: Should we move this logger related changes to a different PR since its not related to s2 storage flags functionality

File: pkg/kopia/cli/repository/storage/s3/arg_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package s3
+
```
Comment (kale-amruta) [inline pkg/kopia/cli/repository/storage/s3/arg_test.go:2]: Is `ArgTest` a utilitity to be used across different packages. This doesnt seem relevant to s3 package. Should we move it to a different package or utils package?

File: pkg/kopia/cli/repository/storage/s3/s3_opts.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	if endpoint == "" {
+		return command.NewNoopArgument()
+	}
+	return command.NewOptionWithArgument("--endpoint", endpoint)
+}
+
+// optPrefix creates a new prefix option with a given prefix.
+// If the prefix is empty, the prefix option is not set.
+func optPrefix(prefix string) command.Applier {
+	return command.NewOptionWithArgument("--prefix", prefix)
```
Comment (kale-amruta) [inline pkg/kopia/cli/repository/storage/s3/s3_opts.go:48]: Why aren`t we returning `command.NewNoopArgument()` over here if `prefix` is empty , the way we are doing over [here](https://github.com/kanisterio/kanister/pull/2659/files#diff-c8a5d40489b59da0012ce94134e6325909185be70772d76a8caba9dbf98ee7a8R55)

File: pkg/kopia/cli/repository/storage/s3/s3_opts_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	{
+		Name:        "optDisableTLS",
+		Argument:    command.NewArguments(optDisableTLS(true), optDisableTLS(false)),
+		ExpectedCLI: []string{"cmd", "--disable-tls"},
+	},
+	{
+		Name:        "optDisableTLSVerify",
+		Argument:    command.NewArguments(optDisableTLSVerify(true), optDisableTLSVerify(false)),
+		ExpectedCLI: []string{"cmd", "--disable-tls-verification"},
+	},
```
Comment (kale-amruta) [inline pkg/kopia/cli/repository/storage/s3/s3_opts_test.go:59]: Can we add test for `optRegion` as well

**Review by kale-amruta — COMMENTED** [review]
left few comments

File: pkg/kopia/cli/repository/storage/s3/s3_opts.go (hunk 3 above)
Comment (plar) [reply pkg/kopia/cli/repository/storage/s3/s3_opts.go:48]: In Kopia, `--prefix=` is a valid option, and our [current implementation](https://github.com/kanisterio/kanister/blob/f65bc3ca1ad1ea035a1ee793e7b011076d5877ac/pkg/kopia/command/storage/s3_args.go#L49) implements the same logic, we don't want to change it.

File: pkg/kopia/cli/repository/storage/s3/arg_test.go (hunk 2 above)
Comment (plar) [reply pkg/kopia/cli/repository/storage/s3/arg_test.go:2]: Agreed. I will move it to [pkg/kopia/cli/internal/test](https://github.com/kanisterio/kanister/blob/be4031d37595569244239a5068af77470b83ca97/pkg/kopia/cli/internal/test/)

File: pkg/kopia/cli/repository/storage/s3/s3_opts_test.go (hunk 4 above)
Comment (plar) [reply pkg/kopia/cli/repository/storage/s3/s3_opts_test.go:59]: Ouch, good catch. :) I will.

File: pkg/kopia/cli/internal/log/log.go (hunk 1 above)
Comment (plar) [reply pkg/kopia/cli/internal/log/log.go:74]: I included the `StringLogger` here primarily because it's key for testing and debugging S3 storage flags. 
We already have 12 PRs in our PR train, adding another one just for the logger could complicate things further. 

Keeping related changes together, like this logger that directly supports S3 storage flags, allows for smoother integration and testing. It seems more efficient to manage this way. 

Happy to discuss this further if there are any concerns!

**Review by kale-amruta — APPROVED** [review]
lgtm
